# -*- coding: utf-8 -*-

import numpy as np
from PyQt5 import QtCore, QtGui, QtWidgets
from matplotlib.backends.backend_qt5agg import FigureCanvas, NavigationToolbar2QT
from matplotlib.figure import Figure
from matplotlib.ticker import EngFormatter

class TableModel(QtCore.QAbstractTableModel):

    def __init__(self, data):
        super(TableModel, self).__init__()
        self._data = data

    def data(self, index, role):
        if role == QtCore.Qt.DisplayRole:
            value = self._data.iloc[index.row(), index.column()]
            return str(value)

    def rowCount(self, index):
        return self._data.shape[0]

    def columnCount(self, index):
        return self._data.shape[1]

    def headerData(self, section, orientation, role):
        # section is the index of the column/row.
        if role == QtCore.Qt.DisplayRole:
            if orientation == QtCore.Qt.Horizontal:
                return str(self._data.columns[section])

            if orientation == QtCore.Qt.Vertical:
                return str(self._data.index[section])

class MplWidget(QtWidgets.QWidget):
    
    def __init__(self, parent = None):

        QtWidgets.QWidget.__init__(self, parent)
        
        self.canvas = FigureCanvas(Figure(tight_layout=True))
        self.mpl_toolbar = NavigationToolbar2QT(self.canvas, parent)

        vertical_layout = QtWidgets.QVBoxLayout()
        vertical_layout.addWidget(self.canvas)
        vertical_layout.addWidget(self.mpl_toolbar)

        self.setLayout(vertical_layout)
        
        
        self.canvas.axes = self.canvas.figure.add_subplot(111)
        self.setup_axes()
        # self.canvas.axes.xaxis.set_ticks([20,100,1000,10000,20000])
        
    def setup_axes(self):
        self.canvas.axes.set_title('Aislamiento de ruido aéreo de un panel simple')
        self.canvas.axes.set_xscale('log')
        self.canvas.axes.grid(which='both')
        formatter0 = EngFormatter(unit='Hz',sep="\N{THIN SPACE}")
        formatter1 = EngFormatter(unit='dB',sep="\N{THIN SPACE}")
        self.canvas.axes.xaxis.set_major_formatter(formatter0)        
        self.canvas.axes.yaxis.set_major_formatter(formatter1)        
        self.canvas.axes.set_xlabel('Frecuencias')
        self.canvas.axes.set_ylabel('Aislamiento')
        self.canvas.axes.set_ylim(0,100)
        self.canvas.axes.set_xlim(20,20000)
        self.canvas.axes.set_xticks(np.array([31.5,63,125,250,500,
                                              1000,2000,4000,8000,16000]))

class Ui_MainWindow(object):
    def setupUi(self, MainWindow):
        MainWindow.setObjectName("MainWindow")
        MainWindow.resize(800, 600)
        sizePolicy = QtWidgets.QSizePolicy(QtWidgets.QSizePolicy.Preferred, QtWidgets.QSizePolicy.Preferred)
        sizePolicy.setHorizontalStretch(0)
        sizePolicy.setVerticalStretch(0)
        sizePolicy.setHeightForWidth(MainWindow.sizePolicy().hasHeightForWidth())
        MainWindow.setSizePolicy(sizePolicy)
        MainWindow.setMinimumSize(QtCore.QSize(800, 600))
        palette = QtGui.QPalette()
        brush = QtGui.QBrush(QtGui.QColor(255, 255, 255))
        brush.setStyle(QtCore.Qt.SolidPattern)
        palette.setBrush(QtGui.QPalette.Active, QtGui.QPalette.WindowText, brush)
        brush = QtGui.QBrush(QtGui.QColor(255, 255, 255))
        brush.setStyle(QtCore.Qt.SolidPattern)
        palette.setBrush(QtGui.QPalette.Active, QtGui.QPalette.Base, brush)
        brush = QtGui.QBrush(QtGui.QColor(0, 0, 0))
        brush.setStyle(QtCore.Qt.SolidPattern)
        palette.setBrush(QtGui.QPalette.Active, QtGui.QPalette.Window, brush)
        brush = QtGui.QBrush(QtGui.QColor(255, 255, 255))
        brush.setStyle(QtCore.Qt.SolidPattern)
        palette.setBrush(QtGui.QPalette.Inactive, QtGui.QPalette.WindowText, brush)
        brush = QtGui.QBrush(QtGui.QColor(255, 255, 255))
        brush.setStyle(QtCore.Qt.SolidPattern)
        palette.setBrush(QtGui.QPalette.Inactive, QtGui.QPalette.Base, brush)
        brush = QtGui.QBrush(QtGui.QColor(0, 0, 0))
        brush.setStyle(QtCore.Qt.SolidPattern)
        palette.setBrush(QtGui.QPalette.Inactive, QtGui.QPalette.Window, brush)
        brush = QtGui.QBrush(QtGui.QColor(190, 190, 190))
        brush.setStyle(QtCore.Qt.SolidPattern)
        palette.setBrush(QtGui.QPalette.Disabled, QtGui.QPalette.WindowText, brush)
        brush = QtGui.QBrush(QtGui.QColor(0, 0, 0))
        brush.setStyle(QtCore.Qt.SolidPattern)
        palette.setBrush(QtGui.QPalette.Disabled, QtGui.QPalette.Base, brush)
        brush = QtGui.QBrush(QtGui.QColor(0, 0, 0))
        brush.setStyle(QtCore.Qt.SolidPattern)
        palette.setBrush(QtGui.QPalette.Disabled, QtGui.QPalette.Window, brush)
        MainWindow.setPalette(palette)
        font = QtGui.QFont()
        font.setFamily("Sawasdee")
        font.setPointSize(10)
        font.setBold(True)
        font.setItalic(True)
        font.setWeight(75)
        MainWindow.setFont(font)
        self.mainwidget = QtWidgets.QWidget(MainWindow)
        self.mainwidget.setMinimumSize(QtCore.QSize(798, 0))
        self.mainwidget.setObjectName("mainwidget")
        self.gridLayout_3 = QtWidgets.QGridLayout(self.mainwidget)
        self.gridLayout_3.setObjectName("gridLayout_3")
        self.input_frame = QtWidgets.QGridLayout()
        self.input_frame.setObjectName("input_frame")
        self.label_title = QtWidgets.QLabel(self.mainwidget)
        self.label_title.setObjectName("label_title")
        self.input_frame.addWidget(self.label_title, 0, 0, 1, 1)
        self.label_title_2 = QtWidgets.QLabel(self.mainwidget)
        self.label_title_2.setObjectName("label_title_2")
        self.input_frame.addWidget(self.label_title_2, 0, 1, 1, 2)
        self.select_layout = QtWidgets.QVBoxLayout()
        self.select_layout.setObjectName("select_layout")
        self.checkbox_cremer = QtWidgets.QCheckBox(self.mainwidget)
        self.checkbox_cremer.setObjectName("checkbox_cremer")
        self.select_layout.addWidget(self.checkbox_cremer)
        self.checkbox_sharp = QtWidgets.QCheckBox(self.mainwidget)
        self.checkbox_sharp.setObjectName("checkbox_sharp")
        self.select_layout.addWidget(self.checkbox_sharp)
        self.checkbox_iso = QtWidgets.QCheckBox(self.mainwidget)
        self.checkbox_iso.setObjectName("checkbox_iso")
        self.select_layout.addWidget(self.checkbox_iso)
        self.checkbox_davy = QtWidgets.QCheckBox(self.mainwidget)
        self.checkbox_davy.setObjectName("checkbox_davy")
        self.select_layout.addWidget(self.checkbox_davy)
        self.input_frame.addLayout(self.select_layout, 1, 1, 1, 1)
        self.process_layout = QtWidgets.QVBoxLayout()
        self.process_layout.setObjectName("process_layout")
        self.button_process = QtWidgets.QPushButton(self.mainwidget)
        self.button_process.setObjectName("button_process")
        self.process_layout.addWidget(self.button_process)
        self.button_clear = QtWidgets.QPushButton(self.mainwidget)
        self.button_clear.setObjectName("button_clear")
        self.process_layout.addWidget(self.button_clear)
        self.input_frame.addLayout(self.process_layout, 1, 2, 1, 1)
        self.input_layout = QtWidgets.QGridLayout()
        self.input_layout.setObjectName("input_layout")
        self.label_thick = QtWidgets.QLabel(self.mainwidget)
        self.label_thick.setObjectName("label_thick")
        self.input_layout.addWidget(self.label_thick, 1, 0, 1, 1)
        self.label_filt = QtWidgets.QLabel(self.mainwidget)
        self.label_filt.setObjectName("label_filt")
        self.input_layout.addWidget(self.label_filt, 3, 0, 1, 1)
        self.label_unit_2 = QtWidgets.QLabel(self.mainwidget)
        self.label_unit_2.setObjectName("label_unit_2")
        self.input_layout.addWidget(self.label_unit_2, 2, 2, 1, 1)
        self.input_lx = QtWidgets.QLineEdit(self.mainwidget)
        self.input_lx.setObjectName("input_lx")
        self.input_layout.addWidget(self.input_lx, 2, 1, 1, 1)
        self.label_unit = QtWidgets.QLabel(self.mainwidget)
        self.label_unit.setObjectName("label_unit")
        self.input_layout.addWidget(self.label_unit, 2, 6, 1, 1)
        self.label_lx = QtWidgets.QLabel(self.mainwidget)
        self.label_lx.setObjectName("label_lx")
        self.input_layout.addWidget(self.label_lx, 2, 0, 1, 1)
        self.label_unit_1 = QtWidgets.QLabel(self.mainwidget)
        self.label_unit_1.setObjectName("label_unit_1")
        self.input_layout.addWidget(self.label_unit_1, 1, 2, 1, 1)
        self.input_thick = QtWidgets.QLineEdit(self.mainwidget)
        self.input_thick.setObjectName("input_thick")
        self.input_layout.addWidget(self.input_thick, 1, 1, 1, 1)
        self.cb_material = QtWidgets.QComboBox(self.mainwidget)
        self.cb_material.setObjectName("cb_material")
        self.input_layout.addWidget(self.cb_material, 0, 1, 1, 5)
        self.label_ly = QtWidgets.QLabel(self.mainwidget)
        self.label_ly.setObjectName("label_ly")
        self.input_layout.addWidget(self.label_ly, 2, 3, 1, 1)
        self.radio_octave = QtWidgets.QRadioButton(self.mainwidget)
        self.radio_octave.setObjectName("radio_octave")
        self.input_layout.addWidget(self.radio_octave, 3, 4, 1, 1)
        spacerItem = QtWidgets.QSpacerItem(250, 20, QtWidgets.QSizePolicy.Expanding, QtWidgets.QSizePolicy.Minimum)
        self.input_layout.addItem(spacerItem, 0, 6, 1, 1)
        self.label_material = QtWidgets.QLabel(self.mainwidget)
        self.label_material.setObjectName("label_material")
        self.input_layout.addWidget(self.label_material, 0, 0, 1, 1)
        self.input_ly = QtWidgets.QLineEdit(self.mainwidget)
        self.input_ly.setObjectName("input_ly")
        self.input_layout.addWidget(self.input_ly, 2, 4, 1, 1)
        self.radio_third = QtWidgets.QRadioButton(self.mainwidget)
        self.radio_third.setObjectName("radio_third")
        self.input_layout.addWidget(self.radio_third, 3, 1, 1, 1)
        self.input_frame.addLayout(self.input_layout, 1, 0, 1, 1)
        self.gridLayout_3.addLayout(self.input_frame, 0, 0, 1, 1)
        self.output_tabs = QtWidgets.QTabWidget(self.mainwidget)
        palette = QtGui.QPalette()
        brush = QtGui.QBrush(QtGui.QColor(0, 0, 0))
        brush.setStyle(QtCore.Qt.SolidPattern)
        palette.setBrush(QtGui.QPalette.Active, QtGui.QPalette.WindowText, brush)
        brush = QtGui.QBrush(QtGui.QColor(255, 255, 255))
        brush.setStyle(QtCore.Qt.SolidPattern)
        palette.setBrush(QtGui.QPalette.Active, QtGui.QPalette.Base, brush)
        brush = QtGui.QBrush(QtGui.QColor(203, 197, 193))
        brush.setStyle(QtCore.Qt.SolidPattern)
        palette.setBrush(QtGui.QPalette.Active, QtGui.QPalette.Window, brush)
        brush = QtGui.QBrush(QtGui.QColor(0, 0, 0))
        brush.setStyle(QtCore.Qt.SolidPattern)
        palette.setBrush(QtGui.QPalette.Inactive, QtGui.QPalette.WindowText, brush)
        brush = QtGui.QBrush(QtGui.QColor(255, 255, 255))
        brush.setStyle(QtCore.Qt.SolidPattern)
        palette.setBrush(QtGui.QPalette.Inactive, QtGui.QPalette.Base, brush)
        brush = QtGui.QBrush(QtGui.QColor(203, 197, 193))
        brush.setStyle(QtCore.Qt.SolidPattern)
        palette.setBrush(QtGui.QPalette.Inactive, QtGui.QPalette.Window, brush)
        brush = QtGui.QBrush(QtGui.QColor(190, 190, 190))
        brush.setStyle(QtCore.Qt.SolidPattern)
        palette.setBrush(QtGui.QPalette.Disabled, QtGui.QPalette.WindowText, brush)
        brush = QtGui.QBrush(QtGui.QColor(203, 197, 193))
        brush.setStyle(QtCore.Qt.SolidPattern)
        palette.setBrush(QtGui.QPalette.Disabled, QtGui.QPalette.Base, brush)
        brush = QtGui.QBrush(QtGui.QColor(203, 197, 193))
        brush.setStyle(QtCore.Qt.SolidPattern)
        palette.setBrush(QtGui.QPalette.Disabled, QtGui.QPalette.Window, brush)
        self.output_tabs.setPalette(palette)
        self.output_tabs.setObjectName("output_tabs")
        self.plot_tab = QtWidgets.QWidget()
        self.plot_tab.setObjectName("plot_tab")
        self.MplWidget = MplWidget(self.plot_tab)
        self.MplWidget.setGeometry(QtCore.QRect(9, 10, 761, 351))
        self.MplWidget.setObjectName("MplWidget")
        self.output_tabs.addTab(self.plot_tab, "")
        self.table_tab = QtWidgets.QWidget()
        self.table_tab.setObjectName("table_tab")
        self.gridLayout_2 = QtWidgets.QGridLayout(self.table_tab)
        self.gridLayout_2.setObjectName("gridLayout_2")
        self.show_fc = QtWidgets.QLineEdit(self.table_tab)
        self.show_fc.setReadOnly(True)
        self.show_fc.setObjectName("show_fc")
        self.gridLayout_2.addWidget(self.show_fc, 1, 1, 1, 1)
        self.table_R = QtWidgets.QTableView(self.table_tab)
        self.table_R.setObjectName("table_R")
        self.gridLayout_2.addWidget(self.table_R, 0, 0, 1, 5)
        self.label_fc = QtWidgets.QLabel(self.table_tab)
        self.label_fc.setObjectName("label_fc")
        self.gridLayout_2.addWidget(self.label_fc, 1, 0, 1, 1)
        self.label_unit_3 = QtWidgets.QLabel(self.table_tab)
        self.label_unit_3.setObjectName("label_unit_3")
        self.gridLayout_2.addWidget(self.label_unit_3, 1, 2, 1, 1)
        self.button_export = QtWidgets.QPushButton(self.table_tab)
        self.button_export.setObjectName("button_export")
        self.gridLayout_2.addWidget(self.button_export, 1, 4, 1, 1)
        self.label_fd = QtWidgets.QLabel(self.table_tab)
        self.label_fd.setObjectName("label_fd")
        self.gridLayout_2.addWidget(self.label_fd, 2, 0, 1, 1)
        self.label_unit_4 = QtWidgets.QLabel(self.table_tab)
        self.label_unit_4.setObjectName("label_unit_4")
        self.gridLayout_2.addWidget(self.label_unit_4, 2, 2, 1, 1)
        self.show_fd = QtWidgets.QLineEdit(self.table_tab)
        self.show_fd.setReadOnly(True)
        self.show_fd.setObjectName("show_fd")
        self.gridLayout_2.addWidget(self.show_fd, 2, 1, 1, 1)
        spacerItem1 = QtWidgets.QSpacerItem(400, 20, QtWidgets.QSizePolicy.Expanding, QtWidgets.QSizePolicy.Minimum)
        self.gridLayout_2.addItem(spacerItem1, 1, 3, 1, 1)
        self.output_tabs.addTab(self.table_tab, "")
        self.gridLayout_3.addWidget(self.output_tabs, 1, 0, 1, 1)
        MainWindow.setCentralWidget(self.mainwidget)
        self.statusbar = QtWidgets.QStatusBar(MainWindow)
        self.statusbar.setObjectName("statusbar")
        MainWindow.setStatusBar(self.statusbar)

        self.retranslateUi(MainWindow)
        self.output_tabs.setCurrentIndex(0)
        QtCore.QMetaObject.connectSlotsByName(MainWindow)

    def retranslateUi(self, MainWindow):
        _translate = QtCore.QCoreApplication.translate
        MainWindow.setWindowTitle(_translate("MainWindow", "TP1 - Acustica 2"))
        self.label_title.setText(_translate("MainWindow", "DATOS DE ENTRADA"))
        self.label_title_2.setText(_translate("MainWindow", "MÉTODOS DE CÁLCULO"))
        self.checkbox_cremer.setText(_translate("MainWindow", "Cremer"))
        self.checkbox_sharp.setText(_translate("MainWindow", "Sharp"))
        self.checkbox_iso.setText(_translate("MainWindow", "ISO 12354-1"))
        self.checkbox_davy.setText(_translate("MainWindow", "Davy"))
        self.button_process.setText(_translate("MainWindow", "Procesar"))
        self.button_clear.setText(_translate("MainWindow", "Borrar"))
        self.label_thick.setText(_translate("MainWindow", "Espesor:"))
        self.label_filt.setText(_translate("MainWindow", "Filtrado:"))
        self.label_unit_2.setText(_translate("MainWindow", "m"))
        self.label_unit.setText(_translate("MainWindow", "m"))
        self.label_lx.setText(_translate("MainWindow", "Largo:"))
        self.label_unit_1.setText(_translate("MainWindow", "m"))
        self.label_ly.setText(_translate("MainWindow", "Alto:"))
        self.radio_octave.setText(_translate("MainWindow", "Octava"))
        self.label_material.setText(_translate("MainWindow", "Material:"))
        self.radio_third.setText(_translate("MainWindow", "Tercio de octava"))
        self.output_tabs.setTabText(self.output_tabs.indexOf(self.plot_tab), _translate("MainWindow", "Grafico"))
        self.label_fc.setText(_translate("MainWindow", "Frecuencia Critica:"))
        self.label_unit_3.setText(_translate("MainWindow", "Hz"))
        self.button_export.setText(_translate("MainWindow", "Exportar..."))
        self.label_fd.setText(_translate("MainWindow", "Frecuencia de Densidad:"))
        self.label_unit_4.setText(_translate("MainWindow", "Hz"))
        self.output_tabs.setTabText(self.output_tabs.indexOf(self.table_tab), _translate("MainWindow", "Tablas"))


# if __name__ == "__main__":
#     app = QtWidgets.QApplication(sys.argv)
#     MainWindow = QtWidgets.QMainWindow()
#     ui = Ui_MainWindow()
#     ui.setupUi(MainWindow)
#     MainWindow.show()
#     sys.exit(app.exec_())

